samcli/hook_packages/terraform/hooks.py

```diff
@@ -0,0 +1,357 @@
+        tf_prop_value = tf_properties.get(tf_prop_name)
+        if tf_prop_value is not None:
+            code[cfn_prop_name] = tf_prop_value
+    return code
+
+
+AWS_LAMBDA_FUNCTION_PROPERTY_BUILDER_MAPPING: PropertyBuilderMapping = {
+    "FunctionName": _get_property_extractor("function_name"),
+    "Architectures": _get_property_extractor("architectures"),
+    "Environment": _build_lambda_function_environment_property,
+    "Code": _build_lambda_function_code_property,
+    "Handler": _get_property_extractor("handler"),
+    "PackageType": _get_property_extractor("package_type"),
+    "Runtime": _get_property_extractor("runtime"),
+    "Layers": _get_property_extractor("layers"),
+    "Timeout": _get_property_extractor("timeout"),
+}
+
+RESOURCE_TRANSLATOR_MAPPING: Dict[str, ResourceTranslator] = {
+    TF_AWS_LAMBDA_FUNCTION: ResourceTranslator(CFN_AWS_LAMBDA_FUNCTION, AWS_LAMBDA_FUNCTION_PROPERTY_BUILDER_MAPPING),
+}
+
+
+def _build_cfn_logical_id(tf_address: str) -> str:
+    """
+    Builds a CloudFormation logical ID out of a terraform resource address
+
+    Parameters
+    ----------
+    tf_address: str
+        terraform resource address
+
+    Returns
+    -------
+    str
+        CloudFormation logical ID
+    """
+    # ignores non-alphanumericals, makes uppercase the first alphanumerical char and the
+    # alphanumerical char right after a non-alphanumerical char
+    chars: List[str] = []
+    nextCharUppercase = True
+    for char in tf_address:
+        if len(chars) == LOGICAL_ID_MAX_HUMAN_LEN:
+            break
+        if not char.isalnum():
+            nextCharUppercase = True
+            continue
+        if nextCharUppercase:
+            chars.append(char.upper())
+            nextCharUppercase = False
+        else:
+            chars.append(char)
+
+    # Add a hash to avoid naming conflicts
+    human_part = "".join(chars)
+    hash_part = str_checksum(tf_address)[:LOGICAL_ID_HASH_LEN].upper()
+
+    return human_part + hash_part
+
+
+def _get_s3_object_hash(bucket: str, key: str) -> str:
+    """
+    Creates a hash for an AWS S3 object out of the bucket and key
+
+    Parameters
+    ----------
+    bucket: str
+        bucket for the S3 object
+    key: str
+        key for the S3 object
+
+    Returns
+    -------
+    str
+        hash for the given bucket and key
+    """
+    md5 = hashlib.md5()
+    md5.update(bucket.encode())
+    md5.update(key.encode())
+    return md5.hexdigest()
+
+
+def _map_s3_sources_to_functions(s3_hash_to_source: Dict[str, str], cfn_resources: Dict[str, Any]) -> None:
+    """
+    Maps the source property of terraform AWS S3 object resources into the the Code property of
+    CloudFormation AWS Lambda Function resources
+
+    Parameters
+    ----------
+    s3_hash_to_source: Dict[str, str]
+        Mapping of S3 object hash to S3 object source
+    cfn_resources: dict
+        CloudFormation resources
+    """
+    for _, resource in cfn_resources.items():
```

Comment:
can we add some debugging message here to let us know that for a lambda resource X, we replaced the S3 properties with some S3 Object file path?

Reply:
done